### PR TITLE
Add UDOP

### PR DESCRIPTION
**Pull request by zinengtang**

### CODE REVIEW

**Comment by zinengtang** [comment]
@microsoft-github-policy-service agree [company="{your company}"]

**Comment by zinengtang** [comment]
@microsoft-github-policy-service agree

File: i-Code-Doc/README.md [hunk 1]
```diff
@@ -1,9 +1,53 @@
-# Unifying Vision, Text, and Layout for Universal Document Processing
+# [Unifying Vision, Text, and Layout for Universal Document Processing](https://arxiv.org/pdf/2212.02623)
+[Zineng Tang](https://zinengtang.github.io/),
+[Ziyi Yang](https://www.microsoft.com/en-us/research/people/ziyiyang/),
```
Comment (ziyi-yang) [inline i-Code-Doc/README.md:3]: Could you change my webpage to https://ziyi-yang.github.io/ ? Thanks!

File: i-Code-Doc/README.md [hunk 2]
```diff
@@ -1,9 +1,53 @@
...
+- [x] Release Model (Encoder + Text decoder)
+- [ ] Release Scripts
+- [ ] Vision Decoder / Weights (Due to fake document generation ethical consideration: we don't have plan on this part yet)
```
Comment (ziyi-yang) [inline i-Code-Doc/README.md:17]: Maybe say: "Due to fake document generation ethical consideration, we plan to release this functionality as an Azure API."

File: i-Code-Doc/assets/_ [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
Comment (ziyi-yang) [inline i-Code-Doc/assets/_:1]: what this file is for? It seems empty.

File: i-Code-Doc/core/common/utils.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            Normalize(
+            mean=[0.485, 0.456, 0.406],
+            std=[0.229, 0.224, 0.225]
+        )])
+
+    image = trans(image)  # copy to make it writeable
+    return image
+
+
+def img_trans_torchvision_int(image, image_size=384):
```
Comment (ziyi-yang) [inline i-Code-Doc/core/common/utils.py:117]: This function seems exactly the same as the previous one, except for the func name.

File: i-Code-Doc/core/common/utils.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+from PIL import Image
+import torchvision.transforms as T
+from torchvision.transforms import functional as F
+
+logger = logging.getLogger(__name__)
+PREFIX_CHECKPOINT_DIR = 'checkpoint'
+_re_checkpoint = re.compile(r'^' + PREFIX_CHECKPOINT_DIR + r'\-(\d+)$')
+    
+    
+def get_visual_bbox(image_size=224):
```
Comment (ziyi-yang) [inline i-Code-Doc/core/common/utils.py:17]: Maybe add several lines of comment on what this function is for.

File: i-Code-Doc/core/models/udop_unimodel.py [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+        else:
+            err_msg_prefix = "decoder_" if self.is_decoder else ""
+            raise ValueError(f"You have to specify either {err_msg_prefix}inputs or {err_msg_prefix}inputs_embeds")
+
+        if inputs_embeds is None:
+            assert self.embed_tokens is not None, "You have to intialize the model with valid token embeddings"
+            inputs_embeds = self.embed_tokens(input_ids)
+                
+        if inputs_patches is not None:
+            #===========================
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_unimodel.py:305]: Maybe we don't need this line? #================

File: i-Code-Doc/core/models/udop_unimodel.py [hunk 7]
```diff
@@ -0,0 +1,678 @@
...
+            # layer_outputs is a tuple with:
+            # hidden-states, key-value-states, (self-attention weights), (self-attention position bias), (cross-attention weights), (cross-attention position bias)
+            if use_cache is False:     # MP fixes
+                layer_outputs = layer_outputs[:1] + (None,) + layer_outputs[1:]
+            hidden_states, present_key_value_state = layer_outputs[:2]
+
+            # We share the position biases between the layers - the first layer store them
+            # layer_outputs = hidden-states, key-value-states (self-attention weights),
+            # (self-attention position bias), (cross-attention weights), (cross-attention position bias)
+            #import pdb; pdb.set_trace()
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_unimodel.py:395]: delete this line

File: i-Code-Doc/core/models/udop_unimodel.py [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+            position_bias = layer_outputs[2]
+            if self.is_decoder and encoder_hidden_states is not None:
+                encoder_decoder_position_bias = layer_outputs[4 if output_attentions else 3]
+            # append next layer key value states
+            if use_cache:
+                present_key_value_states = present_key_value_states + (present_key_value_state,)
+
+            #import pdb; pdb.set_trace()
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_unimodel.py:403]: delete this line

File: i-Code-Doc/core/models/udop_unimodel.py [hunk 9]
```diff
@@ -0,0 +1,678 @@
...
+            input_ids = input_ids.view(-1, input_shape[-1])
+        elif inputs_embeds is None and input_ids is not None and torch.numel(input_ids) == 0:
+            input_ids = torch.full((4, 1024), self.config.pad_token_id, device=input_ids.device, dtype=input_ids.dtype)
+            attention_mask = torch.zeros((4, 1024), device=input_ids.device, dtype=input_ids.dtype)
+            seg_data = torch.zeros((4, 1024, 4), device=input_ids.device, dtype=input_ids.dtype)
+            input_shape = input_ids.size()
+            position_bias = torch.zeros_like(
+                self.get_extended_attention_mask(attention_mask, input_shape, attention_mask.device)
+            )
+            # encoder_attention_mask = attention_mask
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_unimodel.py:292]: delete this line.

File: i-Code-Doc/core/models/udop_dual.py [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+            input_ids = input_ids.view(-1, input_shape[-1])
+        elif inputs_embeds is None and input_ids is not None and torch.numel(input_ids) == 0:
+            input_ids = torch.full((4, 1024), self.config.pad_token_id, device=input_ids.device, dtype=input_ids.dtype)
+            attention_mask = torch.zeros((4, 1024), device=input_ids.device, dtype=input_ids.dtype)
+            seg_data['tokens']['bboxes'] = torch.zeros((4, 1024, 4), device=input_ids.device, dtype=input_ids.dtype)
+            input_shape = input_ids.size()
+            position_bias = torch.zeros_like(
+                self.get_extended_attention_mask(attention_mask, input_shape, attention_mask.device)
+            )
+            # encoder_attention_mask = attention_mask
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_dual.py:180]: Pls delete this line

File: i-Code-Doc/core/models/udop_dual.py [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+            encoder_seq_length = encoder_hidden_states.shape[1]
+            encoder_attention_mask = torch.ones(
+                batch_size, encoder_seq_length, device=inputs_embeds.device, dtype=torch.long
+            )
+
+        # initialize past_key_values with `None` if past does not exist
+        if past_key_values is None:
+            past_key_values = [None] * len(self.block)
+
+        # ourselves in which case we just need to make it broadcastable to all heads.
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_dual.py:213]: not sure if I understand what "ourselves" mean here.

File: i-Code-Doc/core/models/udop_dual.py [hunk 12]
```diff
@@ -0,0 +1,465 @@
...
+            # (self-attention position bias), (cross-attention weights), (cross-attention position bias)
+            #import pdb; pdb.set_trace()
+            position_bias = layer_outputs[2]
+            if self.is_decoder and encoder_hidden_states is not None:
+                encoder_decoder_position_bias = layer_outputs[4 if output_attentions else 3]
+            # append next layer key value states
+            if use_cache:
+                present_key_value_states = present_key_value_states + (present_key_value_state,)
+
+            #import pdb; pdb.set_trace()
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_dual.py:270]: delete

**Review by ziyi-yang — COMMENTED** [review]
Thank you! I've made a few comments.

File: i-Code-Doc/core/models/udop_dual.py [hunk 13]
```diff
@@ -257,18 +249,13 @@ def forward(
                 layer_outputs = layer_outputs[:1] + (None,) + layer_outputs[1:]
             hidden_states, present_key_value_state = layer_outputs[:2]
             # We share the position biases between the layers - the first layer store them
-            # layer_outputs = hidden-states, key-value-states (self-attention weights),
-            # (self-attention position bias), (cross-attention weights), (cross-attention position bias)
+
             #import pdb; pdb.set_trace()
```
Comment (ziyi-yang) [inline i-Code-Doc/core/models/udop_dual.py:253]: Pls double check all file and ensure #import pdb; pdb.set_trace() are all deleted.

File: i-Code-Doc/scripts/finetune_rvlcdip.sh [hunk 14]
```diff
@@ -0,0 +1,23 @@
+CUDA_VISIBLE_DEVICES=0,1,2,3 PYTHONPATH='.' python run_rvlcdip.py \
+--data_dir "./dataset/rvl-cdip" \
+--max_seq_length 1024 \
+--image_size 1024 \
```
Comment (ziyi-yang) [inline i-Code-Doc/scripts/finetune_rvlcdip.sh:4]: image size 224

**Review by ziyi-yang — APPROVED** [review]
Approved.

**Pull request merged**